pkg/resources/grant_privileges_to_share.go

```diff
@@ -0,0 +1,474 @@
+			if err := d.Set("view_name", id.Identifier.FullyQualifiedName()); err != nil {
+				return nil, err
+			}
+		}
+
+		return []*schema.ResourceData{d}, nil
+	}
+}
+
+func CreateGrantPrivilegesToShare(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
+	db := meta.(*sql.DB)
+	client := sdk.NewClientFromDB(db)
+	id := createGrantPrivilegesToShareIdFromSchema(d)
+	log.Printf("[DEBUG] created identifier from schema: %s", id.String())
+
+	err := client.Grants.GrantPrivilegeToShare(ctx, getObjectPrivilegesFromSchema(d), getShareGrantOn(d), sdk.NewAccountObjectIdentifier(id.ShareName.Name()))
+	if err != nil {
+		return diag.Diagnostics{
+			diag.Diagnostic{
+				Severity: diag.Error,
+				Summary:  "An error occurred when granting privileges to share",
+				Detail:   fmt.Sprintf("Id: %s\nShare name: %s\nError: %s", id.String(), id.ShareName, err.Error()),
+			},
+		}
+	}
+
+	d.SetId(id.String())
+
+	return ReadGrantPrivilegesToShare(ctx, d, meta)
+}
+
+func UpdateGrantPrivilegesToShare(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
+	db := meta.(*sql.DB)
+	client := sdk.NewClientFromDB(db)
+
+	id, err := ParseGrantPrivilegesToShareId(d.Id())
+	if err != nil {
+		return diag.Diagnostics{
+			diag.Diagnostic{
+				Severity: diag.Error,
+				Summary:  "Failed to parse internal identifier",
+				Detail:   fmt.Sprintf("Id: %s\nError: %s", d.Id(), err.Error()),
+			},
+		}
+	}
+
+	if d.HasChange("privileges") {
+		before, after := d.GetChange("privileges")
+		privilegesBeforeChange := expandStringList(before.(*schema.Set).List())
+		privilegesAfterChange := expandStringList(after.(*schema.Set).List())
+
+		var privilegesToAdd, privilegesToRemove []sdk.ObjectPrivilege
+
+		for _, privilegeBeforeChange := range privilegesBeforeChange {
+			if !slices.Contains(privilegesAfterChange, privilegeBeforeChange) {
+				privilegesToRemove = append(privilegesToRemove, sdk.ObjectPrivilege(privilegeBeforeChange))
+			}
+		}
+
+		for _, privilegeAfterChange := range privilegesAfterChange {
+			if !slices.Contains(privilegesBeforeChange, privilegeAfterChange) {
+				privilegesToAdd = append(privilegesToAdd, sdk.ObjectPrivilege(privilegeAfterChange))
+			}
+		}
+
+		grantOn := getShareGrantOn(d)
+
+		if len(privilegesToAdd) > 0 {
+			err = client.Grants.GrantPrivilegeToShare(
+				ctx,
+				privilegesToAdd,
+				grantOn,
+				sdk.NewAccountObjectIdentifier(id.ShareName.Name()),
+			)
+			if err != nil {
+				return diag.Diagnostics{
+					diag.Diagnostic{
+						Severity: diag.Error,
+						Summary:  "Failed to grant added privileges",
+						Detail:   fmt.Sprintf("Id: %s\nPrivileges to add: %v\nError: %s", d.Id(), privilegesToAdd, err.Error()),
+					},
+				}
+			}
+		}
+
+		if len(privilegesToRemove) > 0 {
+			err = client.Grants.RevokePrivilegeFromShare(
+				ctx,
+				privilegesToRemove,
+				grantOn,
+				sdk.NewAccountObjectIdentifier(id.ShareName.Name()),
+			)
+			if err != nil {
+				return diag.Diagnostics{
```

Comment:
so the point of having diags is so you can append warnings and such. but here it seems you are always just returning on diagnostic. is this intentional?

Reply:
Yes it is, I just didn't find a case in this resource to append a diag to a slice and return it later on. Most of the diags created here are Errors that should be returned immediately. The only Warning is when show on all tables in schemas should happen, but I also returned, because for now, we don't support that case (like in the other grant resources).